lib/puppetx/simp/compliance_map.rb

```diff
@@ -237,167 +231,199 @@ def write_server_report(config, report)
-          {"value" => data["value"], "identifiers" => data["identifiers"]}
-        end
-        parameters = {}
-        classes = {}
-        @catalog.resources.each do |obj|
-          classname = obj.title.downcase
-          obj.parameters.each do |parameter, data|
-            fully_qualified_parameter = classname + "::" + parameter.to_s
-            if (profile_map.key?(fully_qualified_parameter))
-              profile_settings = profile_map[fully_qualified_parameter]
-              current_value = data.value
-              # XXX ToDo This should be improved to allow for validators to be specified
-              # instead of forcing regexes to be in values (as it breaks enforcement)
-              # ie, functions or built ins.
-              if (profile_settings.key?("value"))
-                expected_value = profile_settings["value"]
-                result = {
-                    "compliant_value" => expected_value,
-                    "system_value" => current_value,
-                }
-                if (profile_settings.key?("identifiers"))
-                 result["identifiers"] = profile_settings["identifiers"]
-                end
-                classkey = "#{obj.type}[#{obj.title}]"
-                if (expected_value =~ /^re:(.+)/)
-                  if (current_value =~ Regexp.new($1))
-                    section = "compliant"
-                  else
-                    section = "non_compliant"
-                  end
-                else
-                  if (current_value == expected_value)
-                    section = "compliant"
-                  else
-                    section = "non_compliant"
-                  end
-                end
-                if report_types.include?(section)
-                  unless (profile_report[section].key?(classkey))
-                    profile_report[section][classkey] = {}
-                    profile_report[section][classkey]["parameters"] = {}
-                  end
-                    profile_report[section][classkey]["parameters"][parameter.to_s] = result
-
-                end
-              end
-            else
-               parameters[fully_qualified_parameter] = true
-               classes[classname] = true
+  report_types = main_config[:report_types]
+
+  if report_types.include?('full')
+    report_types << 'unknown_resources'
+    report_types << 'unknown_parameters'
+    report_types << 'compliant'
+    report_types << 'non_compliant'
+    report_types << 'custom_entries'
+  end
+
+  report = main_config[:extra_data].dup
+
+  report['version'] = '1.0.1';
+  report['timestamp'] = Time.now.to_s
+  report['compliance_profiles'] = {}
+
+  profile_list = get_compliance_profiles
+
+  Array(profile_list).each do |profile|
+    profile_report = {}
+
+    if main_config[:custom_call]
+      add_custom_entries(main_config)
+    else
+      profile_map = profile_compiler.list_puppet_params([profile]).cook do |fully_qualified_parameter, data|
```

Comment:
You are changing cook's api and method signature and breaking all consumers of the library.

There are two better options:
* Add puppet-parameter to the data hash on v1 so it is available via a well defined index
* use list_puppet_params which already yields puppet-parameter and data, and store the result.

The former should be pretty simple to add

Reply:
This seems to be my only issue, since most of these changes are confined to the compliance_map function and are performance or bug fix focused (mark-1-eyeball)

Reply:
@heliocentric Thanks for that, it let me cut out the remaining redundant loop! PR update in a few after the tests pass.